src/main/java/seedu/address/logic/commands/EditCommand.java

```diff
@@ -105,10 +113,37 @@ private static Person createEditedPerson(Person personToEdit, EditPersonDescript
         Address updatedAddress = editPersonDescriptor.getAddress().orElse(personToEdit.getAddress());
         Course updatedCourse = editPersonDescriptor.getCourse().orElse(personToEdit.getCourse());
         Tag updatedTag = editPersonDescriptor.getTag().orElse(personToEdit.getTag());
+
         ArrayList<Module> updatedModules = editPersonDescriptor.getModules().orElse(personToEdit.getModules());
+        if (editPersonDescriptor.hasModuleChanges()) {
+            Module oldModule = editPersonDescriptor.oldModule;
+            Module newModule = editPersonDescriptor.newModule;
+
+            boolean isModuleRenamed = false;
+            if (updatedModules.stream().anyMatch(m -> m.value.equals(newModule.value))) {
+                throw new CommandException(EditCommand.MESSAGE_DUPLICATE_MODULE);
+            }
+
+            for (int i = 0; i < updatedModules.size(); i++) {
+                if (updatedModules.get(i).value.equals(oldModule.value)) {
+                    Module updatedModule = new Module(newModule.value);
+                    if (updatedModules.get(i).hasGrade()) {
+                        updatedModule.setGrade(updatedModules.get(i).getGrade());
+                    }
+                    isModuleRenamed = true;
+                    updatedModules.set(i, updatedModule);
+                    break;
+                }
+            }
+
+            if (!isModuleRenamed) {
+                throw new CommandException(MESSAGE_MODULE_NOT_FOUND);
```

Comment:
Exception handling is appropriately managed, providing meaningful feedback and improving user experience.